### ISSUE TITLE
[Android] Respect custom system-wide text selection toolbar buttons

### ISSUE THREAD

**Issue by bleroux** [flutter/flutter#139361]
### Edit

> This feature is implemented. It is directly available in newly created Flutter project since Flutter 3.19.
> For older project, the following section should be added to the Android manifest file:
> 
> ```xml
>  
> <queries>
>   <intent>
>     <action android:name="android.intent.action.PROCESS_TEXT"/>
>     <data android:mimeType="text/plain"/>
>   </intent>
> </queries>
> ```

### Use case

On Android, it's possible to write an app that adds a custom button to the system-wide text selection toolbar.  For example, here the Android app AnkiDroid has added the "Anki Card" button to my text selection toolbar, and it appears in any app, such as the Google app here.

<img width="300" src="https://user-images.githubusercontent.com/389558/178827259-299e765b-e58c-4233-9bd4-367400058abe.png" />

### Proposal

Flutter should respect these added buttons, so that in any Flutter app running on Android, the buttons that the user has added to their device show up in the default text selection toolbar just like in any native app.

This is the Android part of https://github.com/flutter/flutter/issues/107603 which is broad (all plaforms).
Creating this issue to make it easier to follow.

### Implementation steps

| # | Status | Description | Related Work |
| --- | --- | --- | --- | 
1| Done| Engine - Add Android support for querying and invoking text processing actions.| https://github.com/flutter/engine/pull/44579|
2| Done| Framework - Add ProcessTextService to intereact with the engine | https://github.com/flutter/flutter/pull/137145|
3| Done| fluttter_tools - Add queries section to Android manifest file | https://github.com/flutter/flutter/pull/137207|
4| Done| Framework - Add queried text processing actions to the text selection toolbar for `EditableText` | https://github.com/flutter/flutter/pull/139738|
5| Done| Framework - Add queried text processing actions to the text selection toolbar for `SelectionRegion`| https://github.com/flutter/flutter/pull/141103|

**Comment by gnprice** [comment]
Great to see this happening!

From trying out current main including #139738, though, I think it hasn't yet fully implemented this issue.

Specifically, the new toolbar buttons appear when I select some text in an editable field, but AFAICT they don't appear when I select text elsewhere — in particular text in `Text` widgets inside a `SelectionArea`.

Both cases are about the text selection toolbar. For many of the custom system-wide text selection toolbar buttons, the case of read-only text is by far the most important: for example, all three that appear in the screenshot above (Translate, Anki Card, and Pleco) are primarily for consuming text that other people wrote and that is therefore much more likely to be found in a read-only text widget.

**Comment by gnprice** [comment]
Separately: in an existing app, the new feature didn't work at all when I first tried it on an upgraded Flutter. I fixed it by remembering that we'd discussed the need for a new element in the manifest, looking through the list of PRs above, finding the code that was added to the template manifest, then copy-pasting that into my app's manifest.

In those circumstances, I think it's likely that most existing apps will never pick up this change.

Unfortunately I don't think I have a great immediate suggestion for how to fix that, so we may just live with it (and be happy for all the future new apps, and for existing apps whose developers do notice the lack of these buttons and discover what to add to the manifest to fix it).

* The ideal thing would probably be a mechanism like the one @Hixie mentioned at https://github.com/flutter/flutter/issues/107603#issuecomment-1678210979 :
  > IIRC we have logic to automatically add the NETWORK permission during debug mode, we might be able to automatically add this too?

  But I just looked, and it doesn't seem like we currently have such a mechanism. Instead, the template puts `<uses-permission android:name="android.permission.INTERNET"/>` in the files `android/app/src/{debug,profile}/AndroidManifest.xml`. (But not `…/src/main/…`, which is why it doesn't affect release builds.) So that's the analogue of the solution already implemented in #137207.

  We could build such a mechanism, and in fact I think that'd probably be a good idea. The crux of it would just be to provide one more manifest source file for the Android build process to [merge together](https://developer.android.com/build/manage-manifests#merge-manifests). This is something Flutter plugins already routinely do. Once we do, we could use it for the INTERNET permission (thereby simplifying the templates for newly-created apps), and could also choose to use it for this (with the same benefit plus delivering this feature to apps that predate it).

  Seems beyond the scope of this issue, though.

* Potentially we could also write a migration to add this to the manifest. But if for some reason a developer wanted to opt out of it, I'm not sure that approach would have a great answer for them. (By contrast, when merging manifests, the Android build tools have [a well-defined mechanism](https://developer.android.com/build/manage-manifests#merge_rule_markers) for the app's own source manifest to override or exclude specific elements from libraries' source manifests.)

* We could try to document this more visibly in hopes more developers see it. But it's not a breaking change, so doesn't properly fit in the list of [migration guides](https://docs.flutter.dev/release/breaking-changes). And I kind of doubt many developers read those proactively anyway — their main value is after a developer has already run into a breaking change and is figuring out what to do about it.

**Comment by justinmc** [comment]
@camsim99 Do we have a way to change the Android manifest file for existing Flutter apps like this (see last comment)?

**Comment by bleroux** [comment]
Back from vacation, I wish you all an happy new year! :sparkles: 

@gnprice 
Based on your findings, I filed https://github.com/flutter/flutter/pull/141103 to make the custom buttons visible inside `SelectionArea` toolbar.

In https://github.com/flutter/flutter/pull/139738 I was focused on `TextField` and also quickly checked that it was working for `SelectableText`, it worked there because `SelectableText` relies on `EditableText`, that's not the case for `SelectionArea`).

**Comment by camsim99** [comment]
> @camsim99 Do we have a way to change the Android manifest file for existing Flutter apps like this (see last comment)?

The options listed in [that comment](https://github.com/flutter/flutter/issues/139361#issuecomment-1878206806) all seem like feasible solutions :) @gnprice can you point me towards an example where plugins merge manifest files? That certainly seems safer than writing our own migrator, so I'm curious about what it looks like.

**Comment by gnprice** [comment]
Sure. If you run the following command from the root of the flutter/packages repo, it'll show you all the examples in those plugins:
`git ls-files '*/android/src/*/AndroidManifest.xml' | xargs head -20`

Here's one of them:
```
==> packages/local_auth/local_auth_android/android/src/main/AndroidManifest.xml <==
<manifest xmlns:android="http://schemas.android.com/apk/res/android"
  package="io.flutter.plugins.localauth">
    <uses-permission android:name="android.permission.USE_BIOMETRIC" />
</manifest>
```

So that causes a line `<uses-permission android:name="android.permission.USE_BIOMETRIC" />` to appear in the output manifest for an app that uses that plugin.

---

You can also take an app that you have lying around, open up its `android/` subdirectory as its own project in Android Studio, open the manifest file, and look at [the "Merged Manifest" view](https://developer.android.com/build/manage-manifests#inspect_the_merged_manifest_and_find_conflicts). That will show you the output manifest that ultimately goes into the app, and identify all the different manifest source files that contributed to it. For example in [my app](https://github.com/zulip/zulip-flutter) that screen looks like this:
![image](https://github.com/flutter/flutter/assets/28173/1bea6b82-a62f-4f05-b673-dbcafe46b6f3)
and those "android.*" sources in the right-hand pane are mostly Flutter plugins, including some from flutter/packages and some from elsewhere.

**Comment by bleroux** [comment]
Unassigning myself because I probably not the right person to work on the remaining part (finding a way to automatically update existing manifest files).

**Comment by piedcipher** [comment]
I'm removing the fixed label as issue is still open.

**Comment by bleroux** [comment]
Closing this issue as it is fixed since Flutter 3.19.
For Flutter projects created since Flutter 3.19, the feature is automatically available.
For projects created before Flutter 3.19, a section should be added to the Android manifest. I edited the issue description accordingly and I will open another issue to track the following question: "Do we have a way to change the Android manifest file for existing Flutter apps like this (see last comment)?"
